openedx_events/avro_attrs_bridge.py

```diff
@@ -0,0 +1,287 @@
+        "type": "org.openedx.test.test.test.v0",
+    }
+
+    def __init__(self, attrs_cls, extensions=None, config=None):
+        """
+        Init method for Avro Attrs Bridge
+
+        Arguments:
+            attrs_cls: Attr Class Object (not instance)
+            extensions: dict mapping Class Object to its AvroAttrsBridgeExtention subclass instance
+            config: dict with followings keys
+                - source:  This field will be used to indicate the logical source of an event, and will be of the form
+                           /{namespace}/{service}/{web|worker}.
+                           All services in standard distribution of Open edX should use openedx for the namespace.
+                           Examples of services might be “discovery”, “lms”, “studio”, etc.
+                           The value “web” will be used for events emitted by the web application,
+                           and “worker” will be used for events emitted by asynchronous tasks such as celery workers.
+                           For more, see: https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0041-arch-async-server-event-messaging.html#id3  # pylint: disable=line-too-long
+                - sourcehost: should represent the physical source of message
+                              -- i.e. host identifier of the server that emitted this event (example: edx.devstack.lms)
+                - type: The name of event. Should be formatted `{Reverse DNS}.{Architecture Subdomain}.{Subject}.{Action}.{Major Version}`.
+                        To find out more see: https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0041-arch-async-server-event-messaging.html#id5
+        """
+        self._attrs_cls = attrs_cls
+
+        self.extensions = {}
+        self.extensions.update(self.default_extensions)
+        if isinstance(extensions, dict):
+            self.extensions.update(extensions)
+
+        self.config = {}
+        self.config.update(self.default_config)
+        if isinstance(config, dict):
+            self.config.update(config)
+
+        # Used by record_field_for_attrs_class function to keep track of which
+        # records have already been defined in schema.
+        # Reason: fastavro does no allow you to define record with same name twice
+        self.schema_record_names = set()
+        self.schema_dict = self._attrs_to_avro_schema(attrs_cls)
+        # make sure the schema is parsable
+        fastavro.parse_schema(self.schema_dict)
+
+    def schema_str(self):
+        return json.dumps(self.schema_dict, sort_keys=True)
+
+    def _attrs_to_avro_schema(self, attrs_cls):
+        """
+        Generate avro schema for attr_cls
+
+        Arguments:
+            attrs_cls: Attr class object
+        Returns:
+            complex dict that defines avro schema for attrs_cls
+        """
+        base_schema = {
+            "namespace": "io.cloudevents",
```

Comment:
1. I just found https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md. Have you read this? For example, see `ce_` prefix notes. I'm not clear on when this comes into play.
2. It looks like `time` is optional. Should we leave out any optional attributes if we are unsure? See https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#optional-attributes
3. Is `type` supposed to be something like: https://github.com/eduNEXT/openedx-events/blob/2fba7ae850ae5e0cb6f56026ec17d867bac23d94/openedx_events/learning/signals.py#L19? Are these Signal classes supposed to be for internal/external events (even though Signal is in the name), or was the intention to create new classes for these events with a similar type?
4. Are we using CloudEvent types for all event data? For example, would timestamps be handled correctly? See https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#type-system
5. Do we have any clear use case of how/where using the CloudEvent spec will be helpful? Is there a tool or anything we can use to validate that we are in fact doing what we need to be doing? I found this validator, but without a specific use case, I'm not sure that we can ensure we are doing what we need to be doing (see https://www.itb.ec.europa.eu/json/cloudevents/upload)
6. Can you add a comment with a link to the spec and to the OEP to clarify our use of CloudEvents?

Some of this may require discussion with eduNEXT and Dave O. This may require iterating. Hopefully you can help manage what is good enough and what next steps (tickets) are needed. Thanks!

Reply:
This is interesting. 

For 1. I don't know where kafka-protocol-bindings would come into play for us. So far, I've only dealt with how cloudevents work with avro format.

2. I added time based on specification in [OEP-41](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0041-arch-async-server-event-messaging.html).  How editable is that OEP and how much can we deviate from it for our work?
3.  Yes, I think the type is suppose to be like that. I don't know much about those signal classes.
4. I think we are using cloudevent types for time correctly. We'll be outputting timestamps in ISO format.
5. I haven't yet found a validator for cloundevents specification.
6. I will comment with link to spec and OEP soon.